### DISCUSSION TITLE
Make it easier to select America/New_York for the timezone

### DISCUSSION THREAD

**Discussion by dan-robertson** [grafana/grafana#68965] [General]
The title of this issue is the specific thing that I would like to be improved but there are several ways to do it and some of those may be more or less useful in other cases. I’m happy to change this request to be more general, e.g. ‘Improve timezone selector’.

Here is what usually happens when I try to set the timezone to America/New_york:
1. First I need to navigate tot he timezone selector. This means:
   1. Clicking the time-range selector
   2. Clicking ‘Change time settings’
   3. Clicking the select box
2. Now I can type to search. Generally the issue is that many of the things I try don’t lead to useful results:
   - `EST` either gives too many options to choose from or won’t show up America/New_York in the summer when `EDT` must be typed instead
   - `nyc` doesn’t work: we don’t have the information that NYC is an acceptable abbreviation for New York
   - `new y` (or any longer prefix of ‘new york’) won’t work because the search is strict about the distinction between spaces and underscores
   - `new_y` will find it.

To some extent, I guess I should just learn to be better at knowing the right thing to type, but I still think some improvements could be made that would be useful to all users of the feature.

The most obvious improvement to me is to make searching that field a little less sensitive to the difference between underscores and spaces. The function which does the search comes from [`useFilterBySearchIndex`](https://github.com/grafana/grafana/blob/2ceed6a16971c02ecfa94dcd79c680b958e67395/packages/grafana-ui/src/components/DateTimePickers/TimeZonePicker.tsx#L153-L160). It could be changed to tread words of the search term separately, something like this:

```diff
 const useFilterBySearchIndex = () => {
+  let cachedQuery = "";
+  let cachedSplit = [];
   return useCallback((option: SelectableValue, searchQuery: string) => {
     if (!searchQuery || !option.data || !option.data.searchIndex) {
       return true;
     }
+    if (cachedQuery !== searchQuery) {
+        cachedQuery = searchQuery;
+        cachedSplit = searchQuery.split(/\s+/).filter(word => word !== "").map(toLower)
+    }

-    return option.data.searchIndex.indexOf(toLower(searchQuery)) > -1;
+    return cachedSplit.every(word => option.data.searchIndex.indexOf(word) > -1);
   }, []);
 };
```

Another option would be converting the search term into a RegExp, quoting as appropriate and replacing spaces with `[ _]` or `.*`. That requires that the ordering of words in queries matches the ordering in the result, and maybe that makes them more predictable for users.

The above ideas would help with my specific problem of it being hard to find America/New_York. Here are some other ideas to improving the timezone selector, at least for my usecases:
- Remember recent selections and make them easier to choose. I think I only use about 4 timezones. There are a few ways to do this (e.g. ranking instead of just filtering the dropdown, offering those recents at the top of the dropdown)
- Reduce the number of clicks. For example, focus the timezone dropdown when clicking ‘Change time settings’. Maybe ‘Change time settings’ should really be two options – changing the timezone and choosing when the fiscal year starts seem like quite different options; maybe the latter should be a date rather than time setting.
- Have some configurable way to have a few easy-to-reach-for buttons for specific timezones (e.g. the same four timezones I would like to choose between are shared by every user of grafana)
- Have some way to narrow down the list of timezones searched so that it is easy to find the relevant ones.
- Have more data go into the search index (e.g. if you search for ‘San Francisco’ you won’t find anything – you need to know to search for Los_Angeles instead; similarly, ‘nyc’ won’t find New_York)

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). We are migrating this discussion to an issue and closing the discussion. The issue is #82704. Feel free to continue the discussion around this there. Thank you!
